### ISSUE TITLE
[Request] Expanded inv/cube/stash

### ISSUE THREAD

**Issue by Dugahst** [mapron/d2modgen#17]
not a bug, but a request - can you add an option to expand the inventory, cube, and stash size?

**Comment by Dugahst** [comment]
I understand, can you maybe post instructions on how to use the 'pre-gen data' feature?

**Comment by mapron** [comment]
![image](https://github.com/mapron/d2modgen/assets/7624327/a0ffaf97-9f31-4094-9650-d76fd5c347f8)
something like this I suppose? just select mod there

**Comment by mapron** [comment]
@KaBombz  What mod you trying to use and what merge strategy?

**Comment by KaBombz** [comment]
Tried with both Vanilla++ and ReMoDDeD with all merge strategies, the error occurs regardless.

**Comment by mapron** [comment]
@KaBombz  no, Bonesy mods are binary, so unfortunately they won't work, I am sorry.

**Comment by KaBombz** [comment]
Ah, okay, was confused by the screenshot provided above in that case

**Comment by mapron** [comment]
Sorry for confusion, I have no valid mod installed on my PC like extended stash. I need to register on nexus and install D2RMM to get this, I personally don't have time for that. If you have standalone extended stash mod I can check if it works right now. People on discrord reported before that it works fine, that's why I recommended pregen step.
Don't want do false advertisement here

**Comment by Lethlman** [comment]
I get this error with this mod. Is there another stash expansion mod that works?
![image](https://github.com/mapron/d2modgen/assets/101858300/ea62e053-b00c-47e4-ace6-382c966629b6)

**Comment by mapron** [comment]
Can you post a link to this mod or something? how do I obtain it for testing?

**Comment by Lethlman** [comment]
I tried these 2 mods
https://www.nexusmods.com/diablo2resurrected/mods/13
https://www.nexusmods.com/diablo2resurrected/mods/377?tab=files
the second one, when i try to generate the mod window just closes, no error.

**Comment by mapron** [comment]
Ok, I will create nexus account, download it and look what I can do. Don't expect fast resolution :(

**Comment by Lethlman** [comment]
> Ok, I will create nexus account, download it and look what I can do. Don't expect fast resolution :(

its all good man. I appreciate the work you put in already and this is truly a fun mod

**Comment by mapron** [comment]
Thank you for the patience. I really confused as other people reported it worked (before) so I was relying on that. Did not had time to test merge mod functionality thoroughly before.

**Comment by Lethlman** [comment]
ya man, this really is just a QOL thing for me.

**Comment by Lethlman** [comment]
> I also thinking about implementing it myself with some png slicing on the fly from diablo2 resurrected data. That would be super cool to implement. And you could choose several inventory sizes, turn on and off mercenary slots etc. Though probably won't be artistically pleasant with auto slicing, need to think about it.

That would be dope. honestly the stash size is the main thing keeping me from playing it more. whatever you can do is awesome.

**Comment by mapron** [comment]
Though it is not an easy task to do, I still keep this issue open, as it is a thing I really want myself

**Comment by bhavinamin** [comment]
Temp workaround -  I have [this ](https://www.nexusmods.com/diablo2resurrected/mods/842)Expanded Stash mod working just fine with modgen + D2RMM. Save your modgen config before doing anything and I recommend testing this with a super minimal mod config to avoid any surprise crashes and then slowly enable what you want. I always do this~!

Steps to reproduce: 

1. Download and setup D2RMM to point to your install folder (I have it set to Steam install). Note if using Steam, you'll need the optional Casclib.dll file in the mod download section and overrite the one in `<D2RMM directory>\Tools`- this is in testing and hasn't been merged into the official release yet. 
2. Generate mod normally without pre-gen and post-gen data. Target Warlock expansion is enabled. 
3. Copy generated `random` directory into your `<D2RMM directory>\mods` . Empty out your actual game directory mods folder just to start clean - this likely isn't necessary but I like to keep things clean. 
4. Move contents from `D2RMM\mods\random\random.mpq\*` to `DRMM\mods\random` (essentially empty out the .mpq folder and delete it after)
5. In D2RMM, you will receive warning that `random` mod is a data mod and to put it at the top of the load order. I made sure mine was at the top of the load order
6. Click Install Mods in D2RMM to generate the `D2RMM\D2RMM.mpq` folders in your actual game directory. 
7. Make sure your launch args is set to `-mod D2RMM -txt`  or `-mod D2RMM -txt -enablerespec`


[Screenshot ](https://imgur.com/a/DpQaKXr)of sanity test with keys in inventory, shorthanded scroll name + expanded stash on a fresh char

**Note: by default D2RMM references its own save file directory, so by default you'll have an empty char list if you've never used it before**

**Comment by mapron** [comment]
@bhavinamin  thank you for an instruction, but I was able to make it work with 'pre-gen' mods and merging feature one day (then things broke). So not like I am anaware how to do that with hand file manipulation. 

It just I want clear solution NOT involving external mod at all (I want something like extracting png, modifying them via some rules, like "cut this region, paste it here" and then saving new UI texture in mod folder).

**Comment by bhavinamin** [comment]
Understood, that is the ideal goal. The findings were mainly for anyone who temporarily wants to the test community mods in addition to randomizing :)

**Comment by mapron** [comment]
@bhavinamin , I got preliminary estimation on that task.  mod.js is pretty clean and understandable. I even thought about adding JS engine just to leave script as is (and provide D2MM callbacks lol) but then decided it is too much. 
I guess I will reimplement that, as I want customizeable inventory fields everywhere... and not just in stash, that only a start!

next, I debugged SpriteEdit program.
https://github.com/eezstreet/D2RModding-SpriteEdit/blob/master/D2RModding-SpriteEdit/Program.cs#L51
looks like files like stashpanelbgshared.sprite use plain RGBA format (rgba data starting from offset 0x28), so I don't need much code added to read-write them (I will use QImage loading for interleaved rgba, plain and simple), just manage a few headers. 

One downside that I want expanded inventory as well... so i just looked source code for
https://www.nexusmods.com/diablo2resurrected/mods/562
And I have a clear understanding what to do.

Yet again, I really hate to repeat others work BUT i need a different approach so just adding JS engine and api from D2MM won't do good for me. 
Overall... I'd say with 20 human-hours this feature is possible. At least for MVP variant, where you don't provide grid size you want. 

Oh, yeah, I know mercenary extra slots ALSO would be nice but that's a feature creep and would be another level (some day... maybe). 

Also I now clearly understand why merging with mods is now should be abandoned; they all now target d2mm with its own js scripting. I'd rather remove pre-gen/post-gen to simplify mogen code.
